README.md
# libchipgpio
Dynamic C Library for interfacing with the GPIO pins on the C.H.I.P. mini computer.
One of the main focuses behind this library is to provide "forwards-compatibility" with possible future revisions of the CHIP. Accessible pins can be added, removed, and modified without breaking compatibility with programs written to run on older revisions of the CHIP.

BUILDING
--------

1. Install essential packages

  `sudo apt-get install build-essential git`

2. Clone this repository

  `git clone https://github.com/BryanHaley/libchipgpio.git`

3. Build and install

  `cd libchipgpio`
  
  `make all`
  
  `sudo make install`
  
4. (Optional) Run the example program

  `sudo ./morse_example`
  
  + Connect an LED to pin XIO-P7 and a button to LCD-VSYNC (with a pull-up resistor) for this example.
  
  `sudo ./toggle_example`
  
  + Connect an LED to pin XIO-P7, a button to XIO-P5, and another button to XIO-P3 for this example.
  
USAGE
-----

###chip_gpio.h

Preamble: this library does require some mindfulness from the coder using it to maintain forward compatibility. Please read *and follow* the best practices section, lest the major advantage of using this library is lost. See `src/example/morse.c` for an implementation.

libchipgpio provides the following functions to interface with the CHIP's GPIO pins. Include `chip_gpio.h` and link `lchipgpio` to use them.

+ `initialize_gpio_interface()`
  
  + Must be called before interfacing with any GPIO pins.
  
  + `init_gpio_inf()`

    + Same as above, except a shorter (less-explicit) name.
  
+ `open_gpio_pin(int pin)`

  + Pins must be opened before they can be read from or written to. This effectively tells the system to begin monitoring the pin. See best practices section for what argument to pass here.
  
+ `set_gpio_dir(int pin, int out)`

  + You must set a direction (input or output) for GPIO pins. Use GPIO_DIR_OUT and GPIO_DIR_IN.
  
+ `setup_gpio_pin(int pin, int out)`

  + Convenience function to call `open_gpio_pin` and `set_gpio_dir` in one line, since it is necessary to do both before a pin can be used.
  
+ `set_gpio_val(int pin, int val`

  + Write a value (1 or 0, GPIO_PIN_ON or GPIO_PIN_OFF, GPIO_PIN_HIGH or GPIO_PIN_LOW) to a pin. The pin must be in the *out* direction to do this. Attempting to do otherwise will cause an error.
  
+ `read_gpio_val(int pin)`

  + Returns the value (1 or 0, GPIO_PIN_ON or GPIO_PIN_OFF, GPIO_PIN_HIGH or GPIO_PIN_LOW) of a pin. The pin must be in the *in* direction to do this. Attempting to do otherwise will cause an error.
  
+ `close_gpio_pin(int pin)`

  + Pins should be closed when no longer in use. This effectively tells the system to stop monitoring the pin. You *can* attempt to close pins before opening them (this will cause a warning) to ensure you will be able to open the pin. However, this isn't recommended, as if the pins aren't closed before start-up it's most likely because you didn't close them properly the last time the program ran, or some other program/script is actively using them.
  
+ `is_gpio_pin_open(int pin)`

  + Returns 1 or 0 based on if the GPIO pin is currently open. Can be used to prevent errors on startup, and/or warn users.
  
+ `get_gpio_pin_num_from_name(char* name)`
  
  + You **must** use this to get the pin number that should be passed to the other functions in this list. Otherwise, forward compatibility will be broken. See the best practices section.

  +  `get_gpio_num(char* name)`
  
    + Same as above, except a shorter (less-explicit) name. Seriously, use one of these.

+ `get_gpio_xio_base()`

  + Not explicitly useful, but if you need the kernel-recognized identifier number for the first XIO pin, you can get it here. This is found based on files that exist in the `/sys/class/gpio` directory, and is kernel-version-agnostic.
  
+ `autoclose_gpio_pins()`

  + You don't *need* to keep track of all the pins you've opened - libchipgpio will do this automatically. You can close all of the pins opened by your program by calling this function.
  
+ `terminate_gpio_interface()`

  + Closes the GPIO file handles and calls autoclose_gpio_pins(). Call this before your program exits, or when you're completely done with GPIO.
  
  + `term_gpio_inf()`
  
    + Same as above, except a shorter (less-explicit) name.
    
+ `_n(char* name` variants
  
  + Functions accepting an `int pin` argument have a variant with the suffix `_n` that accepts a pin name string instead of an integer (e.g. `open_gpio_pin_n("XIO-P0");`). This is compliant with the best practices, however it comes at a performance cost (the pin name string must be compared to a list of pin name strings, which can be costly if done in a loop). Might be useful when grabbing user input or something, but otherwise, stick to the method outlined in the best practices section.
  
###chip_gpio_callback_manager.h

Preamble: Follow the best practices for this interface just as you would the one above it. See `src/example/toggle.c` for an implementation.

To supplement the general libchipgpio interface, `chip_gpio_callback_manager.h` is provided to easily create and manipulate callback functions that are invoked when a pin changes value. The pin values are polled on a separate thread, so that your application may run concurrently.

+ `struct pin_change_t`

  + A simple struct containing two integers, `pin` and `new_val`. A `pin_change_t` variable is passed to a callback function when it is invoked.
  
+ `initialize_callback_manager()`
  
  + Must be called before anything else.
  
  + `init_callback_manager()`
  
    + Same as above, except a shorter (less-explicit) name.
    
+ `start_callback_manager()`

  + Begin polling GPIO pin values. Generally, you'll want to register your callback functions *before* calling this if possible.
  
+ `register_callback_func(int pin, void* func, void* arg)`

  + Any time the GPIO pin `pin` changes value, `func` is invoked and passed a `pin_change_t` and `arg`. Note that `arg` can be a pointer to any data you wish; it is for your use. Keep in mind each GPIO pin can only have one callback function; registering a second callback function to a pin will simply overwrite the original.
  
  + Callback functions should have the following signature: `int foo(pin_change_t, void*)` and return a success or failure code (such as `GPIO_OK` or `GPIO_ERR`).
  
+ `register_callback_flip_func(int pin, void* func, void* arg)`

  + Slight modification on `register_callback_func`. Callbacks registered using this are only invoked when the pin changes value, then changes back to its original value (e.g. rather than invoking the callback function when a button is pressed then again when the button is released, the callback function is invoked only once the button is released after being pressed).
  
  + A GPIO pin is assumed to be in its default state when the callback function is registered. When it changes value, it is in a state of being flipped. Once it changes back to its default state, the callback function is invoked, and the GPIO pin is no longer in a state of being flipped.
  
+ `set_callback_flip_value(int pin, int val)`
  
  + By default, `pin`'s value is read when the callback is registered (which becomes its default state), and the "flip" value is set to the opposite of that. However, this may be undesirable; if for example a pin is connected to a button and the button is pressed down (perhaps by an impatient user) when the callback function is registered, the flip value will be the opposite of what's expected. So, it may be necessary to force a flip value.
  
  + When using buttons, pass one of the following values as an argument for `val`:
    
    + `CALLBACK_ON_PRESS`
      
      + The callback function is invoked when the button is pressed, but not when it is released.
    
    + `CALLBACK_ON_RELEASE`
    
      + The callback function is invoked when the button is released, but not when it is pressed.
      
    + Use the `_PULLDOWN` variants if you are using a pull down resistor instead. Otherwise, it is assumed you are using pull-up resistors, or an XIO pin (which have built-in pull-up resistors).
        
  + When not using buttons, your intentions are more clear when using `GPIO_PIN_HIGH` and `GPIO_PIN_LOW`.
  
+ `remove_callback_func(int pin)`

  + Deregisters the callback function for a pin.
  
+ `pause_callback_manager()`

  + Stop polling the GPIO pins without deregistering all callback functions.
  
+ `unpause_callback_manager()`

  + Resume polling the GPIO pins.
  
+ `terminate_callback_manager()`

  + *MUST* be called when you're finished with callback functions. This deregisters all callback functions and frees up memory. If you need to work with the callback manager again afterwards, it must initialized again.
  
  + `term_callback_manager()`

    + Same as above, except a shorter (less-explicit) name.
    
+ `set_callback_polling_delay(int new_delay)`

  + Optionally, you may set a delay between every 'round' of polling GPIO pins. After polling all pins, the callback manager simply invokes `usleep` with `new_delay` before polling all pins from the beginning again.
  
  + Anything less than 1 is considered no delay.
  
+ `_n(char* name` variants
  
  + Like in the `chip_gpio.h` interface, you may supply a pin's name instead using `_n` variants of any function with the parameter `int pin`.
    
BEST PRACTICES
--------------

+ Use `get_gpio_num(char* name)` to get a GPIO pin's number.

  + **This is by far the most important one to follow.** Supply this function with a pin name string (e.g. "LCD-D4"), and it will return an integer you may use with the above functions (as the `int pin` argument). **Do not directly pass a pin number to the above functions.** In order to maintain forwards compatibility, the possibility of the GPIO pinout changing must be accounted for. If pin numbers are passed directly (e.g. do not do `open_gpio_pin(37);` for LCD-VSYNC) forwards compatibility will probably be broken, as libchipgpio has no way of knowing if Pin 37 will be a usable GPIO pin on future revisions of the CHIP.
  
  + Example of a proper way to setup a pin:
  
    + `int start_button_pin = get_gpio_num("XIO-P7");`
      
      `setup_gpio_pin(start_button_pin, GPIO_DIR_IN);`
    
  + start_button_pin can be reused; you only need to call get_gpio_num once for each pin.
  
  + As of February 2017, the current pinout of the current version of the CHIP is the following:

     ![alt text](https://docs.getchip.com/images/chip_pinouts.jpg)

     + Use these names when accessing the pins. According to the documentation, pins U13 18 to 40 (PWM0 to LCD-DE), U14 13-20 (XIO-P0 to XIO-P7), and U14 27-40 (CSIPCK to CSID7 plus two ground pins) are usable as GPIO. Note however, a few of those are ground pins (and as such are not usable as input/output), and I’m pretty sure pin 17 (LCD-D2) is perfectly acceptable to be used (typo?). There are, effectively, 44 usable GPIO pins at the current time. This is reflected in the current version of the library. Keep in mind if you use an HDMI or VGA adapter, many or all of the non-XIO pins will be unavailable. Use a TRRS->RCA cable or run the CHIP headless with SSH or VNC for projects that use many of GPIO pins.
  
+ Check for errors.

  + Every function in the libchipgpio interface will print information to stderr and return -1 if a major error occurs. Use that to troubleshoot and/or produce meaningful error messages.
  
  + Following the previous example:
  
    + `if (setup_gpio_pin(start_button_pin, GPIO_DIR_IN) < 0)`
      
      `{ fprintf(stderr, "Encountered GPIO error on startup. Shutting down.”); return -1; }`
       
   + libchipgpio functions will generate meaningful error messages on stderr on their own, so it may be more helpful to specify *where* the error occurred than the error itself.
   
   + In particular, monitor `open_gpio_pin` for errors. If you get an error here, check if the pin is already open, prompt the user to allow you to close it (or at least warn them that you will), then try again to open the pin. Keep in mind other programs or scripts may be using GPIO pins, so only close a GPIO pin you didn't open if absolutely necessary.
       
+ Ensure you close the pins by the time your program ends.

  + Leaving a pin open can and will cause errors the next time you open your program. Libchipgpio needs to open pins itself in order to keep track of which pins it can close and to help ensure no other program or script will interfere while your program is running.
  
+ Your user will need this library installed to run your program.

  + This library is dynamically linked so that is may be updated independently of your program (again to help provide forwards compatibility). Therefore, it must be installed on the user's system. I am looking into making a ppa to make this a more streamlined process, but until then, consider automating the cloning, building, and installing of this library as shown in the Building section. Or, at the very least, link to this git repository and inform the user of its necessity.

+ If you need to change the available pins on the CHIP (perhaps a new revision has come out and I have not yet updated the library, for example), edit the `chip_gpio_pin_defs.h` file. Theoretically, it should be possible to make all of the necessary changes by only editing that file (unless some seriously drastic changes have been made to the CHIP pinout). Once you’re done making edits, build and install your new version of the library.
  
TODO
----

+ Add wrappers for interfacing with i2c, SPI, and PWM.
  
  + I'm currently awaiting a shipment containing an i2c/spi LCD screen to test with.

  + In the future, I’ll also look into making expanding GPIO via an i2c GPIO chip as easy as possible (ideally, integrating with the libchipgpio interface as if they were on the CHIP itself). It would be especially useful for those using VGA/HDMI adapters.
  
+ Confirm support for the CHIP Pro.

  + Theoritically it should already be supported, but I have no way of knowing for sure until I get one. There are probably a few pins on the Pro that don't exist on the CHIP, and currently I've only defined the CHIP's pins in `chip_gpio_pin_defs.h`.
  
+ Finished ~~Add a feature utilizing a separate thread to monitor specified GPIO pins to trigger callback functions when a value has changed.~~

  +~~Callbacks on a separate thread would be much more convenient for polling things like buttons and switches.~~
  
BUY ME A COFFEE
---------------

If you found this library useful at all, I'd appreciate it :-)

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/BryanHaley)

OPEN SOURCE LICENSE
-------------------

This software is dual-licensed (GPLv2 and SimplifiedBSD). Choose whichever works best for you.

```
GNU General Public License, version 2

LIBCHIPGPIO - Dynamic library to interface with GPIO pins on the C.H.I.P. mini computer.
Copyright (C) 2017 Bryan Haley

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
```

```
BSD 2-Clause License

Copyright (c) 2017, Bryan Haley
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```


CHANGELOG
CURRENT_VERSION: 1.1.0

Legend:
VERSION NUMBER: MAJOR.MINOR.BUGFIX
    BUGFIX revisions do not add or remove features; only fix bugs.
    MINOR revisions may add or change features, but should not break compatiblity.
    MAJOR revisions may break compatibility.
    Higher importance revisions may contain changes that would otherwise be in a lower priority revision (e.g. a MINOR revision may contain bugfixes or a MAJOR revision may add features or et cetera).

Ver 1.0.0:
* Initial release to Github, library functional.

Ver 1.1.0:
* Added callbacks
* Added "flip" callback (on press or on release)
* Added way to identify pins using a custom function (via function pointer)
* Added toggle function in chip_gpio interface
* Refactored functions (moved variable declarations to the top and some other cleaning)
* Better error messages (print out errno)
* Added forgotten get_gpio_num(char*) to chip_gpio interface
* Fixed pointer issues in initialize_gpio_interface
* Fixed handling of some strings (mainly in chip_gpio_utils.h)
* Removed most magic numbers
* Added new example using callbacks


include/chip_gpio_pin_defs.h
/*
 * Copyright (c) 2017, Bryan Haley
 * This code is dual licensed (GPLv2 and Simplified BSD). Use the license that works
 * best for you. Check LICENSE.GPL and LICENSE.BSD for more details.
 *
 * chip_gpio_pin_defs.h
 * Header file used to define pins in a way we can easily interpret. This file can be
 * edited to add, remove, or modify pins recognized by the library with relative ease.
 * Doing so should not break programs compiled against older versions of the library,
 * unless you remove a pin it relied upon.
 */

#ifndef CHIP_GPIO_PIN_DEFS_H
#define CHIP_GPIO_PIN_DEFS_H

//Probably unwise to change these two
#define GPIO_OK 0
#define GPIO_ERR -1

//Some definitions that detail what the CHIP pinout is like in a way we can interpret

#define FIRST_PIN          1 // First pin is 1 since that's how it's labeled. 0 is
#define NUM_PINS          80 // unused. Keep that in mind when using NUM_PINS.
#define U13_END           40
#define U14_OFFSET        40 //These are the same now, but may not always be.

//If anything has changed with the XIO pins, it's important to reflect those
#define NUM_XIO_U14_PINS   8 //                                 changes here
#define XIO_U14_FIRST_PIN 13 //Use the _ALL variant for pre-baked U14 offset

#define LCD_U13_FIRST_PIN 17
#define LCD_U13_LAST_PIN  40

//Use LCD_U14_(FIRST/LAST)_PIN_ALL for pre-baked U14 offset
#define LCD_U14_FIRST_PIN 27
#define LCD_U14_LAST_PIN  38 //39, 40 is GND

#define GPIO_SYSFS_PATH "/sys/class/gpio/gpio"
#define GPIOCHIP_SYSFS_PATH "/sys/class/gpio/gpiochip"

//If the multiplier of a pin is this, that means it's definitely not an R8 pin
#define GPIO_UNUSED '\0'

// The following are definitions assigning pin numbers to pin names as seen here:
// goo.gl/zpt4ge (for the current version of the CHIP, anyway).
// They are to be used ONLY by libchipgpio internally. Since these are pre-
// processor definitions, they will be baked into whatever includes this file
// and cannot be changed without recompiling. Therefore, if a program uses
// these, it will break forwards-compatibility and the program will need to be
// recompiled if a pinout change breaks it. When best practices are followed,
// only libchipgpio should need to be recompiled.

//U13
#define     GPIO_LCD_D2     17
#define     GPIO_PWM0       18
#define     GPIO_LCD_D4     19
#define     GPIO_LCD_D3     20
#define     GPIO_LCD_D6     21
#define     GPIO_LCD_D5     22
#define     GPIO_LCD_D10    23
#define     GPIO_LCD_D7     24
#define     GPIO_LCD_D12    25
#define     GPIO_LCD_D11    26
#define     GPIO_LCD_D14    27
#define     GPIO_LCD_D13    28
#define     GPIO_LCD_D18    29
#define     GPIO_LCD_D15    30
#define     GPIO_LCD_D20    31
#define     GPIO_LCD_D19    32
#define     GPIO_LCD_D22    33
#define     GPIO_LCD_D21    34
#define     GPIO_LCD_CLK    35
#define     GPIO_LCD_D23    36
#define     GPIO_LCD_VSYNC  37
#define     GPIO_LCD_HSYNC  38
#define     GPIO_LCD_DE     40

//U14
#define     GPIO_XIO_P0     13 + U14_OFFSET
#define     GPIO_XIO_P1     14 + U14_OFFSET
#define     GPIO_XIO_P2     15 + U14_OFFSET
#define     GPIO_XIO_P3     16 + U14_OFFSET
#define     GPIO_XIO_P4     17 + U14_OFFSET
#define     GPIO_XIO_P5     18 + U14_OFFSET
#define     GPIO_XIO_P6     19 + U14_OFFSET
#define     GPIO_XIO_P7     20 + U14_OFFSET

#define	    GPIO_CSIPCK	    27 + U14_OFFSET
#define	    GPIO_CSICK	    28 + U14_OFFSET
#define	    GPIO_CSIHSYNC   29 + U14_OFFSET
#define	    GPIO_CSIVSYNC   30 + U14_OFFSET
#define	    GPIO_CSID0	    31 + U14_OFFSET
#define	    GPIO_CSID1	    32 + U14_OFFSET
#define	    GPIO_CSID2	    33 + U14_OFFSET
#define	    GPIO_CSID3	    34 + U14_OFFSET
#define	    GPIO_CSID4	    35 + U14_OFFSET
#define	    GPIO_CSID5	    36 + U14_OFFSET
#define	    GPIO_CSID6	    37 + U14_OFFSET
#define	    GPIO_CSID7	    38 + U14_OFFSET

char* PIN_UNUSED;
char* XIO_CHIP_LABEL;

//struct containing information to identify a gpio pin
//see: goo.gl/vQLRuW (pages 18 to 20), goo.gl/1cGAZw
typedef struct
{
    char* name;
    char mult;
    int off; 
    //Function pointer so you may find the kernel number of a pin however you please
    void* func;
    void* arg;
    int hard_coded_kern_pin; //DEPRACATED; use function pointer instead
} pin_identifier_t;

// There is a chance the kernel identifier for a pin may be calculated by an algorithm
// not present in this library in the future. To combat this, you may provide a function
// to calculate this yourself. If a function is provided, it takes precendence over any
// over method of finding the kernel number that may apply to it.
// Signature of function pointers to find kernel identifiers:
// int foo(int pin, void* arg);
// The function should return the kernel identifier number. The argument pin is the pin
// number as #defined above. The argument arg is a void pointer you can set yourself.
// Set p_ident[bar].func to the address of your function, and arg to a pointer leading to
// any data you may need. Use a struct if you need arg to contain multiple variables.

pin_identifier_t p_ident[NUM_PINS+FIRST_PIN];

//This is called in initialize_gpio_interface() before anything else
static int initialize_gpio_pin_names()
{
    PIN_UNUSED = "DO NOT USE";
    // this is the i2c chip handling the XIO pins
    XIO_CHIP_LABEL = "pcf8574a"; // Currently, the CHIP uses this label to identify
                                 // the correct directory containing the xio base
                                 // number, so we just need to know this beforehand.

    for (int i = 0; i < NUM_PINS+FIRST_PIN; i++)
    {
	//Unused pins will all have this info
        p_ident[i].name = PIN_UNUSED;
        p_ident[i].mult =  GPIO_UNUSED;
        p_ident[i].off  = GPIO_ERR;
        p_ident[i].func = NULL;
        p_ident[i].arg = NULL;
        p_ident[i].hard_coded_kern_pin = GPIO_ERR;
    }
	
    // Here, we define the name (by which users of this library should access
    // the pins by) and other identifying information. XIO pins should only
    // need the name.
    
    //U13
    p_ident[GPIO_LCD_D2].name        = "LCD-D2";
        p_ident[GPIO_LCD_D2].mult    = 'D'; p_ident[GPIO_LCD_D2].off    =  2;
    p_ident[GPIO_PWM0].name          = "PWM0";
        p_ident[GPIO_PWM0].mult      = 'B'; p_ident[GPIO_PWM0].off      =  2;
    p_ident[GPIO_LCD_D4].name        = "LCD-D4";
        p_ident[GPIO_LCD_D4].mult    = 'D'; p_ident[GPIO_LCD_D4].off    =  4;
    p_ident[GPIO_LCD_D3].name        = "LCD-D3";
        p_ident[GPIO_LCD_D3].mult    = 'D'; p_ident[GPIO_LCD_D3].off    =  3;
    p_ident[GPIO_LCD_D6].name        = "LCD-D6";
        p_ident[GPIO_LCD_D6].mult    = 'D'; p_ident[GPIO_LCD_D6].off    =  6;
    p_ident[GPIO_LCD_D5].name        = "LCD-D5";
        p_ident[GPIO_LCD_D5].mult    = 'D'; p_ident[GPIO_LCD_D5].off    =  5;
    p_ident[GPIO_LCD_D10].name       = "LCD-D10";
        p_ident[GPIO_LCD_D10].mult   = 'D'; p_ident[GPIO_LCD_D10].off   = 10;
    p_ident[GPIO_LCD_D7].name        = "LCD-D7";
        p_ident[GPIO_LCD_D7].mult    = 'D'; p_ident[GPIO_LCD_D7].off    =  7;
    p_ident[GPIO_LCD_D12].name       = "LCD-D12";
        p_ident[GPIO_LCD_D12].mult   = 'D'; p_ident[GPIO_LCD_D12].off   = 12;
    p_ident[GPIO_LCD_D11].name       = "LCD-D11";
        p_ident[GPIO_LCD_D11].mult   = 'D'; p_ident[GPIO_LCD_D11].off   = 11;
    p_ident[GPIO_LCD_D14].name       = "LCD-D14";
        p_ident[GPIO_LCD_D14].mult   = 'D'; p_ident[GPIO_LCD_D14].off   = 14;
    p_ident[GPIO_LCD_D13].name       = "LCD-D13";
        p_ident[GPIO_LCD_D13].mult   = 'D'; p_ident[GPIO_LCD_D13].off   = 13;
    p_ident[GPIO_LCD_D18].name       = "LCD-D18";
        p_ident[GPIO_LCD_D18].mult   = 'D'; p_ident[GPIO_LCD_D18].off   = 18;
    p_ident[GPIO_LCD_D15].name       = "LCD-D15";
        p_ident[GPIO_LCD_D15].mult   = 'D'; p_ident[GPIO_LCD_D15].off   = 15;
    p_ident[GPIO_LCD_D20].name       = "LCD-D20";
        p_ident[GPIO_LCD_D20].mult   = 'D'; p_ident[GPIO_LCD_D20].off   = 20;
    p_ident[GPIO_LCD_D19].name       = "LCD-D19";
        p_ident[GPIO_LCD_D19].mult   = 'D'; p_ident[GPIO_LCD_D19].off   = 19;
    p_ident[GPIO_LCD_D22].name       = "LCD-D22";
        p_ident[GPIO_LCD_D22].mult   = 'D'; p_ident[GPIO_LCD_D22].off   = 22;
    p_ident[GPIO_LCD_D21].name       = "LCD-D21";
        p_ident[GPIO_LCD_D21].mult   = 'D'; p_ident[GPIO_LCD_D21].off   = 21;
    p_ident[GPIO_LCD_CLK].name       = "LCD-CLK";
        p_ident[GPIO_LCD_CLK].mult   = 'D'; p_ident[GPIO_LCD_CLK].off   = 24;
    p_ident[GPIO_LCD_D23].name       = "LCD-D23";
        p_ident[GPIO_LCD_D23].mult   = 'D'; p_ident[GPIO_LCD_D23].off   = 23;
    p_ident[GPIO_LCD_VSYNC].name     = "LCD-VSYNC";
        p_ident[GPIO_LCD_VSYNC].mult = 'D'; p_ident[GPIO_LCD_VSYNC].off = 27;
    p_ident[GPIO_LCD_HSYNC].name     = "LCD-HSYNC";
        p_ident[GPIO_LCD_HSYNC].mult = 'D'; p_ident[GPIO_LCD_HSYNC].off = 26;
    p_ident[GPIO_LCD_DE].name        = "LCD-DE";
        p_ident[GPIO_LCD_DE].mult    = 'D'; p_ident[GPIO_LCD_DE].off    = 25;

    //U14
    p_ident[GPIO_XIO_P0].name = "XIO-P0";
    p_ident[GPIO_XIO_P1].name = "XIO-P1";
    p_ident[GPIO_XIO_P2].name = "XIO-P2";
    p_ident[GPIO_XIO_P3].name = "XIO-P3";
    p_ident[GPIO_XIO_P4].name = "XIO-P4";
    p_ident[GPIO_XIO_P5].name = "XIO-P5";
    p_ident[GPIO_XIO_P6].name = "XIO-P6";
    p_ident[GPIO_XIO_P7].name = "XIO-P7";
	

    p_ident[GPIO_CSIPCK].name       = "CSIPCK";
        p_ident[GPIO_CSIPCK].mult   = 'E'; p_ident[GPIO_CSIPCK].off   =  0;
    p_ident[GPIO_CSICK].name        = "CSICK";
        p_ident[GPIO_CSICK].mult    = 'E'; p_ident[GPIO_CSICK].off    =  1;
    p_ident[GPIO_CSIHSYNC].name     = "CSIHSYNC";
        p_ident[GPIO_CSIHSYNC].mult = 'E'; p_ident[GPIO_CSIHSYNC].off =  2;
    p_ident[GPIO_CSIVSYNC].name     = "CSIVSYNC";
        p_ident[GPIO_CSIVSYNC].mult = 'E'; p_ident[GPIO_CSIVSYNC].off =  3;
    p_ident[GPIO_CSID0].name        = "CSID0";
        p_ident[GPIO_CSID0].mult    = 'E'; p_ident[GPIO_CSID0].off    =  4;
    p_ident[GPIO_CSID1].name        = "CSID1";
        p_ident[GPIO_CSID1].mult    = 'E'; p_ident[GPIO_CSID1].off    =  5;
    p_ident[GPIO_CSID2].name        = "CSID2";
        p_ident[GPIO_CSID2].mult    = 'E'; p_ident[GPIO_CSID2].off    =  6;
    p_ident[GPIO_CSID3].name        = "CSID3";
        p_ident[GPIO_CSID3].mult    = 'E'; p_ident[GPIO_CSID3].off    =  7;
    p_ident[GPIO_CSID4].name        = "CSID4";
        p_ident[GPIO_CSID4].mult    = 'E'; p_ident[GPIO_CSID4].off    =  8;
    p_ident[GPIO_CSID5].name        = "CSID5";
        p_ident[GPIO_CSID5].mult    = 'E'; p_ident[GPIO_CSID5].off    =  9;
    p_ident[GPIO_CSID6].name        = "CSID6";
        p_ident[GPIO_CSID6].mult    = 'E'; p_ident[GPIO_CSID6].off    = 10;
    p_ident[GPIO_CSID7].name        = "CSID7";
        p_ident[GPIO_CSID7].mult    = 'E'; p_ident[GPIO_CSID7].off    = 11;
    
    return 0;
}

#endif
